### PR TITLE
[Collision] Cleaning pass on types

### PR DESCRIPTION
**Pull request by alxbilger**
The goal was to avoid type conversions by removing explicit types. Instead I took the rigorous returned types and added static_assert when two types are interacting.

Curiously, I found out that the templated code (intersection with Sphere) is not called anywhere. Let's see if a known plugin is using it.

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by alxbilger** [comment]
[ci-build][with-all-tests]

**Comment by alxbilger** [comment]
To do: extract the removed code from this PR and create a new one.

File: Sofa/Component/Collision/Detection/Intersection/src/sofa/component/collision/detection/intersection/LocalMinDistance.cpp [hunk 1]
```diff
@@ -586,13 +587,13 @@ int LocalMinDistance::computeIntersection(Triangle& e2, Sphere& e1, OutputVector
 
 bool LocalMinDistance::testIntersection(Line& e2, Point& e1)
 {
-
+    static_assert(std::is_same_v<Line::Coord, Point::Coord>, "Data mismatch");
     if(!e1.isActive(e2.getCollisionModel()) || !e2.isActive(e1.getCollisionModel()))
         return false;
 
     const double alarmDist = getAlarmDistance() + e1.getProximity() + e2.getProximity();
```
Comment (epernod) [inline Sofa/Component/Collision/Detection/Intersection/src/sofa/component/collision/detection/intersection/LocalMinDistance.cpp]:
```suggestion
    const SReal alarmDist = getAlarmDistance() + e1.getProximity() + e2.getProximity();
```

File: Sofa/Component/Collision/Detection/Intersection/src/sofa/component/collision/detection/intersection/MeshDiscreteIntersection.cpp [hunk 2]
```diff
@@ -50,13 +50,17 @@ bool MeshDiscreteIntersection::testIntersection(Triangle&, Line&)
...
-    Vector3 right;
+    static_assert(std::is_same_v<Triangle::Coord, Line::Coord>, "Data mismatch");
+    static_assert(Triangle::Coord::total_size == 3, "Must be a vec type");
+
+    const Triangle::Coord& A = e1.p1();
+    const Triangle::Coord AB = e1.p2()-A;
+    const Triangle::Coord AC = e1.p3()-A;
+    const Line::Coord& P = e2.p1();
+    const Line::Coord PQ = e2.p2()-P;
+    Mat<3, 3, Triangle::Coord::value_type> M;
```
Comment (epernod) [inline Sofa/Component/Collision/Detection/Intersection/src/sofa/component/collision/detection/intersection/MeshDiscreteIntersection.cpp]: NOINIT?

File: Sofa/Component/Collision/Detection/Intersection/src/sofa/component/collision/detection/intersection/MeshMinProximityIntersection.cpp [hunk 3]
```diff
@@ -323,10 +323,10 @@ int MeshMinProximityIntersection::computeIntersection(Triangle& e2, Point& e1, O
 
 bool MeshMinProximityIntersection::testIntersection(Line& e2, Point& e1)
 {
-
+    static_assert(std::is_same_v<Line::Coord, Point::Coord>, "Data mismatch");
     const SReal alarmDist = intersection->getAlarmDistance() + e1.getProximity() + e2.getProximity();
-    const Vector3 AB = e2.p2()-e2.p1();
-    const Vector3 AP = e1.p()-e2.p1();
+    const Point::Coord AB = e2.p2()-e2.p1();
```
Comment (epernod) [inline Sofa/Component/Collision/Detection/Intersection/src/sofa/component/collision/detection/intersection/MeshMinProximityIntersection.cpp]:
```suggestion
    const Line::Coord AB = e2.p2()-e2.p1();
```

**Review by epernod — COMMENTED** [review]
2-3 small details, otherwise ready to go

File: Sofa/Component/Collision/Detection/Intersection/src/sofa/component/collision/detection/intersection/MeshDiscreteIntersection.cpp (hunk 2 above)
Comment (alxbilger) [reply Sofa/Component/Collision/Detection/Intersection/src/sofa/component/collision/detection/intersection/MeshDiscreteIntersection.cpp]: you're right

File: Sofa/Component/Collision/Detection/Intersection/src/sofa/component/collision/detection/intersection/LocalMinDistance.cpp (hunk 1 above)
Comment (alxbilger) [reply Sofa/Component/Collision/Detection/Intersection/src/sofa/component/collision/detection/intersection/LocalMinDistance.cpp]: you're right. And it's not the only one in this file

File: Sofa/Component/Collision/Detection/Intersection/src/sofa/component/collision/detection/intersection/MeshDiscreteIntersection.cpp [hunk 4]
```diff
@@ -50,13 +50,17 @@ bool MeshDiscreteIntersection::testIntersection(Triangle&, Line&)
...
+    static_assert(std::is_same_v<Triangle::Coord, Line::Coord>, "Data mismatch");
+    static_assert(Triangle::Coord::total_size == 3, "Must be a vec type");
+
+    const Triangle::Coord& A = e1.p1();
+    const Triangle::Coord AB = e1.p2()-A;
+    const Triangle::Coord AC = e1.p3()-A;
+    const Line::Coord& P = e2.p1();
+    const Line::Coord PQ = e2.p2()-P;
+    Mat<3, 3, Triangle::Coord::value_type> M(NOINIT);
+    Mat<3, 3, Triangle::Coord::value_type> Minv;
```
Comment (epernod) [inline Sofa/Component/Collision/Detection/Intersection/src/sofa/component/collision/detection/intersection/MeshDiscreteIntersection.cpp]:
```suggestion
    Mat<3, 3, Triangle::Coord::value_type> Minv(NOINIT);
```

**Comment by epernod** [comment]
[ci-build][with-all-tests]

**Pull request merged**